torchmetrics/collections.py

```diff
@@ -182,16 +183,42 @@ def persistent(self, mode: bool = True) -> None:
         """Method for post-init to change if metric states should be saved to
         its state_dict
         """
-        for _, m in self.items():
+        for _, m in self.items(keep_base=True):
             m.persistent(mode)
 
     def _set_name(self, base: str) -> str:
         name = base if self.prefix is None else self.prefix + base
         name = name if self.postfix is None else name + self.postfix
         return name
 
+    def keys(self, keep_base: bool = False):
+        r"""Return an iterable of the ModuleDict key.
+        Args:
+            keep_base: Whether to add prefix/postfix on the items collection.
+        """
+        if keep_base:
+            return self._modules.keys()
+        return OrderedDict(**{self._set_name(k): v for k, v in self._modules.items()}).keys()
```

Comment:
Why do you use an `OrderedDict`? The inner `dict` does not enforce the order so you've lost the order already.

Same below

Reply:
since `MetricCollection` inherent from `nn.ModuleDict` which is ordered dict then this should also be

Reply:
My point is we have:

```python
x = {self._set_name(k): v for k, v in self._modules.items()}
return OrderedDict(x).keys()
```

But the dict comprehension to create `x` loses the order guarantee.

So this should be instead:

```suggestion
        return OrderedDict([(self._set_name(k), v) for k, v in self._modules.items()]).keys()
```

Reply:
Updated.